runtime/src/main/scala/akka/grpc/internal/ChannelUtils.scala

```diff
@@ -88,43 +89,31 @@ object ChannelUtils {
    * INTERNAL API
    */
   @InternalApi
-  private[akka] def monitorChannel2(
+  private[akka] def monitorChannel(
       done: Promise[Done],
       channel: ManagedChannel,
       maxConnectionAttempts: Option[Int]): Unit = {
-    def monitor(connectionAttempts: Int): Unit = {
-      println(s"monitoring with connectionAttempt $connectionAttempts")
+    def monitor(currentState: ConnectivityState, connectionAttempts: Int): Unit = {
+      //TODO: remove before merge
+      println(s"monitoring with state $currentState and connectionAttempts $connectionAttempts")
       if (maxConnectionAttempts.contains(connectionAttempts)) {
         // shutdown is idempotent in ManagedChannelImpl
         channel.shutdown()
         done.tryFailure(new ClientConnectionException(s"Unable to establish connection after [$maxConnectionAttempts]"))
       } else {
-        val currentState = channel.getState(false)
-        println(s"current state $currentState")
-        if (currentState == ConnectivityState.SHUTDOWN) {
-          done.trySuccess(Done)
-        } else {
-          channel.notifyWhenStateChanged(
-            currentState,
-            new Runnable {
-              def run() = {
-                val newState = channel.getState(false)
-                if (newState == ConnectivityState.TRANSIENT_FAILURE) {
-                  monitor(connectionAttempts + 1)
-                } else if (newState == ConnectivityState.READY) {
-                  monitor(0)
-                } else if (newState == ConnectivityState.SHUTDOWN) {
-                  done.trySuccess(Done)
-                } else {
-                  // IDLE / CONNECTING
-                  monitor(connectionAttempts)
-                }
-              }
-            })
-        }
+        channel.notifyWhenStateChanged(currentState, new Runnable {
+          def run(): Unit = channel.getState(false) match {
+            case fail @ ConnectivityState.TRANSIENT_FAILURE => monitor(fail, connectionAttempts + 1)
+            case ready @ ConnectivityState.READY            => monitor(ready, 0)
+            case ConnectivityState.SHUTDOWN                 => done.trySuccess(Done)
```

Comment:
Moving this inside the `notifyWhenStateChanged` seems dangerous when the initial state might already be `SHUTDOWN`. That makes me wonder:
* would it be possible to do the `match` outside of `notifyWhenStateChanged` entirely?
* is it necessary to call `notifyWhenStateChanged` on each recursive call, or would once be sufficient?

Reply:
@raboof If initial state `SHUTDOWN` I complete promise outside `monitor` function and not go into it.
We must to call `notifyWhenStateChanged` each time to track state changes. It call callback only once when passed state would not equal to current anymore.
I think that moving matching outside the `notifyWhenStateChanged` function can be done. I'll try to do it

Reply:
@raboof If initial state `SHUTDOWN` I complete promise outside `monitor` function and not go into it.
We must to call `notifyWhenStateChanged` each time to track state changes. It call callback only once when passed state would not equal to current anymore.
I think that moving matching outside the `notifyWhenStateChanged` function can be done. I'll try to do it